jax/_src/lax/parallel.py

```diff
@@ -586,9 +586,16 @@ def _allreduce_impl(pos_reducer, *args, axes, axis_index_groups):
 
 def _allreduce_abstract_eval(*args, axes, axis_index_groups):
   pos_axes = tuple(axis for axis in axes if isinstance(axis, int))
+  named_shapes = [arg.named_shape for arg in args]
+  if axis_index_groups is None:
+    named_axes = set(axis for axis in axes if not isinstance(axis, int))
+    named_shapes = [{name: size for name, size in arg.named_shape.items()
+                     if name not in named_axes} for arg in args]
```

Comment:
It seems like we should check here that the `named_axes` over which we're reducing are indeed all present in the input. (I'm assuming `psum(1, 'i')` in the "surface language" would generate a `pbroadcast`, maybe via the traceable. An alternative would be to say `psum [ axis_name='i' ] x:int32[]` is well-typed in jaxprs but IIRC we didn't want to go that route. I'm thinking of the abstract eval rule as our typechecking rule, so this is our chance to catch any type errors!)

Reply:
No, I think we do want to go the looser route here. I'm too tired to make the case right now, though :(
(And I disagree with Adam, at least, and probably also the rest of you!)

Reply:
We have since talked extensively about this among the three of us! We haven't figured out exactly what to do, but we have figured out that we can separate concerns: we can land this PR thinking of names as "metadata" (rather than types), in which case it's just tracking more information and not changing any behaviors (or ruling out any future behaviors), and figure out the type-level story separately.